src/dbnode/storage/index/block.go

```diff
@@ -162,32 +471,242 @@ func (b *block) WriteBatch(inserts *WriteBatch) (WriteBatchResult, error) {
 	}
 
 	partialErr, ok := err.(*m3ninxindex.BatchPartialError)
-	if !ok { // should never happen
-		err := b.unknownWriteBatchInvariantError(err)
-		// NB: marking all the inserts as failure, cause we don't know which ones failed
+	if !ok {
+		// NB: marking all the inserts as failure, cause we don't know which ones failed.
 		inserts.MarkUnmarkedEntriesError(err)
 		return WriteBatchResult{NumError: int64(inserts.Len())}, err
 	}
 
 	numErr := len(partialErr.Errs())
 	for _, err := range partialErr.Errs() {
-		// Avoid marking these as success
+		// Avoid marking these as success.
 		inserts.MarkUnmarkedEntryError(err.Err, err.Idx)
 	}
 
-	// mark all non-error inserts success, so we don't repeatedly index them
+	// Mark all non-error inserts success, so we don't repeatedly index them.
 	inserts.MarkUnmarkedEntriesSuccess()
 	return WriteBatchResult{
 		NumSuccess: int64(inserts.Len() - numErr),
 		NumError:   int64(numErr),
 	}, partialErr
 }
 
-func (b *block) executorWithRLock() (search.Executor, error) {
-	var expectedReaders int
-	if b.activeSegment != nil {
-		expectedReaders++
+func (b *block) foregroundCompactWithBuilder(builder segment.DocumentsBuilder) error {
+	// We inserted some documents, need to compact immediately into a
+	// foreground segment.
+	b.Lock()
+	foregroundSegments := b.foregroundSegments
+	b.Unlock()
+
+	segs := make([]compaction.Segment, 0, len(foregroundSegments)+1)
+	segs = append(segs, compaction.Segment{
+		Age:  0,
+		Size: int64(len(builder.Docs())),
+		Type: segments.MutableType,
+	})
+	for _, seg := range foregroundSegments {
+		segs = append(segs, compaction.Segment{
+			Age:     seg.Age(),
+			Size:    seg.Segment().Size(),
+			Type:    segments.FSTType,
+			Segment: seg.Segment(),
+		})
 	}
+
+	plan, err := compaction.NewPlan(segs, b.opts.ForegroundCompactionPlannerOptions())
+	if err != nil {
+		return err
+	}
+
+	// Check plan.
+	var planErr error
+	switch {
+	case len(builder.Docs()) == 0:
+		// Must be compacting with a builder that has some inserted documents.
+		planErr = errForegroundCompactorNoBuilderDocs
+	case len(plan.Tasks) == 0:
+		// Must generate a plan since we gave it a mutable segment type to compact.
+		planErr = errForegroundCompactorNoPlan
+	case !taskHasNilSegment(plan.Tasks[0]):
```

Comment:
I read through `plan.go` and still don't understand this.

Reply:
Yeah so... it's kinda fucked up, so basically the planner thinks everything is always a `segment.Segment` and there's no way for me to put the `Builder` on the `compaction.Segment` struct on the `Segment segment.Segment` field.  

So what I do is set all the other metadata on the `compaction.Segment` struct but leave `Segment segment.Segment` field as `nil`.  It still does the right thing selecting that to be the first one to combine with other segments, but I just basically need to make sure that it always wants to compact it, so I can safely know when I see a `nil` segment I basically need to do a compaction with the builder I have on hand.

I can't really think of a way to do this nicely, possibly I could create a smaller scoped interface to pass to the `compaction` code so the `Builder` could be passed in place there too?  Thoughts?

Reply:
I changed it so it counts the number of "builders" used, should make it a little more simple.